add tests for changes in tsutils

kcpevey opened the pull request:
- add tests for changes in tsutils
- fix minor bugs that the tests exposed


I'd like @CommonClimate or @khider to review the expected values in the tests to ensure that things are working as expected.

MarcoGorelli commented on pyleoclim/tests/test_utils_tsutils.py:
```diff
@@ -0,0 +1,170 @@
+        assert exponent == 0
+        assert direction == 'prograde'
+
+    with pytest.warns(match='Time unit'):
+        (datum, exponent, direction) = tsutils.time_unit_to_datum_exp_dir(time_unit, time_name='unknown')
+        assert direction == 'prograde'
+
+    with pytest.warns(match='Time unit'):
+        (datum, exponent, direction) = tsutils.time_unit_to_datum_exp_dir(time_unit, time_name='age')
+        assert direction == 'retrograde'   
```
why not check `datum` and `exponent` too?

kcpevey replied:
Those are handled in the following tests (there is code following the `time_name` bit that modifies `datum` and `exponent`). This test set is trying to exclusively look at the `time_name` options and implications.

CommonClimate replied:
@khider and I are confused about what this test does. Can you explain the goal?

MarcoGorelli replied:
I think it's just to check what happens when some invalid `time_unit ` is passed

still though, I'd suggest adding
```
        assert datum == 0
        assert exponent == 0
```
, it's only 2 lines of code, makes it consistent with the code above, and the logic inside `time_unit_to_datum_exp_dir` may well change someday such that `time_name` also affects `datum` and `exponent`

kcpevey replied:
The goal of any test suite is to hit 100% coverage. That means we want the test suite to go down every pathway - every if/else, raise etc. 

So this test in particular checks to make sure we get the appropriate warning when 

1. time_unit is some unknown string, time_name is None
2. time_unit is some unknown string, time_name is some unknown string
3. time_unit is some unknown string, time_name is "age"

MarcoGorelli commented on pyleoclim/utils/tsutils.py:
```diff
@@ -48,6 +48,11 @@
 
 SECONDS_PER_YEAR = 365.25 * 60  * 60 * 24
 
+MATCH_A  = ['y', 'yr', 'yrs', 'year', 'years']
+MATCH_KA = ['ky', 'kyr', 'kyrs', 'kiloyear', 'ka'] 
+MATCH_MA = ['ma', 'my','myr','myrs']
+MATCH_GA = ['ga', 'gy', 'gyr', 'gyrs']
```
perhaps make these `frozenset`?

kcpevey replied:
makes sense. will do. 

I'm also not sure:
* how often these get used elsewhere in the codebase (would it be more appropriately placed elsewhere?)
* why `['yr BP', 'yrs BP', 'years BP']` aren't included as a separate list

CommonClimate replied:
hi @kcpevey , we're on it. 
The function tackles a few cases at once, and I am grateful for your review as there might be cases that fall throiugh the cracks. 
> how often these get used elsewhere in the codebase (would it be more appropriately placed elsewhere?)

it's mostly on series creation (`__init__`) and in `convert_time_unit()` (a Series API function), as well as a few other places. I don't think it matters much where they are placed as long as they are accessible. 

> why ['yr BP', 'yrs BP', 'years BP'] aren't included as a separate list

Good question. The issue is that the way most paleo folks report their units is context-dependent and conveys several pieces of information at once. BP = before present 0(which can mean either 1950, 2000, or the date of collection), so it's ambiguous, but always means retrograde time and (~90% of the time) `datum = 1950`.  "years" is a statement about the `exponent` parameter (0 in this case).  However, to complicate matters, the way the variable is called (usually Time or Age) also conveys some information: Age implies retrograde time; Time is more ambiguous.   

Currently the code tries to catch most commonly used idioms, but it will never be 100%, which is why it's necessary to set some sensible defaults, even if the user has to change them after the fact if they didn't get what they wanted.  Feel free to suggest other approaches if you can see a better way to do this.

kcpevey replied:
Thanks, I understand know. The `MATCH_*` statements are specifically about the exponent, but `yr BP` is a mix - BP is about the *datum* while years is about the *exponent*. 

>I don't think it matters much where they are placed as long as they are accessible.

Right. And "accessible" also includes "doesn't cause and import loop" - just something to keep in mind in case it comes up in the future.

MarcoGorelli reviewed (changes requested):
Nice, generally looks really good, just left a couple of minor comments

kcpevey commented on pyleoclim/utils/tsutils.py:
```diff
@@ -58,25 +63,21 @@ def time_unit_to_datum_exp_dir(time_unit, time_name=None):
+    elif time_unit.lower() in MATCH_GA:
         datum = 1950
         exponent = 9
         direction = 'retrograde'
```
I can't comment on the exact lines, but below here on lines 87-95 there is a separate set of if/else statements. Why aren't those included with the ones above?

CommonClimate replied:
```
match_a  = ['y', 'yr', 'yrs', 'year', 'years']
    match_ka = ['ky', 'kyr', 'kyrs', 'kiloyear', 'ka'] 
    match_Ma = ['ma', 'my','myr','myrs']
    match_Ga = ['ga', 'gy', 'gyr', 'gyrs']
```
these are statements about the exponent, which usually tell you about the direction and the datum as well. (eg if you are dealing with Ma or Ga, you could not care less whether the datum is 1950 or 2000).  this bit deals with datum/direction:

```
if 'b2k' in tu:
        datum = 2000
        direction = 'retrograde'
    elif any(c in tu for c in ['bp', 'bnf', 'b1950']):
        datum = 1950
        direction = 'retrograde'
    elif any(c in tu for c in ['ad', 'ce']):
        datum = 0
        direction = 'prograde'
```
It is true that both sets of statements partially overlap, so I can see a potential for some conflict if the users provide highly non-standard names, but I think those rules will work in the vast majority of cases. We will likely have to modify them as we encounter more examples, but that's the bane of being a paleoclimatologist, and we can deal with that, so please leave that on our plate.

kcpevey replied:
Understood. I was just concerned about the overlap. Happy to leave it to you :)

kcpevey commented on pyleoclim/tests/test_utils_tsutils.py:
```diff
@@ -0,0 +1,170 @@
+    assert exponent == 0
+    assert direction == 'prograde'
+
+
```
@CommonClimate @khider What I could really use your help on is to make sure the following 6 tests with `expected` values are correct. Are the expected arrays correct?

CommonClimate replied:
What is `dataframe_dt`? I cannot figure out if the `expected` output is correct because I don't understand what the input is supposed to be.

MarcoGorelli replied:
It's defined in `pyleoclim/tests/conftest.py`:

```
@pytest.fixture
def dataframe_dt():
    """Pandas dataframe with a datetime index and random values"""
    length = 5
    dti = pd.date_range("2018-01-01", periods=length, freq="Y")
    df = pd.DataFrame(np.array(range(length)), index=dti)
    return df
```

It's just a way to create an object which you can re-use in different tests, rather than having to write
```
    length = 5
    dti = pd.date_range("2018-01-01", periods=length, freq="Y")
    dataframe_dt = pd.DataFrame(np.array(range(length)), index=dti)
```
at the beginning of every test

https://youtu.be/ScEQRKwUePI

CommonClimate replied:
Hadn't seen that. Makes a ton of sense, and there are probably of lot of opportunities to improve our existing pytests that way!

MarcoGorelli reviewed (approved):
Looks good to me, over to USC

CommonClimate commented:
This is stretching my Python knowledge a bit (I'd never heard of frozen sets), but it generally looks good. However, I decided to streamline `time_unit_to_datum_exp_dir()` because we were overdoing things. I'm reasonably confident that this version catches all sensible unit specifications (famous last words...) and it removes a couple of tests, which you can just comment out. Can you look at my latest commit and make the changes that appear necessary? 

BTW, @khider pointed out that those conversion functions should be in tsbase.py, not tsutils.py, see #299. When do you think it safest to move them over?

CommonClimate commented:
Specifically, I removed the if/then statement having to do with `match_a`, because we set exponent = 0 by default, so that covers that case. Also, the case of "years BP" is already covered by `elif any(c in tu for c in ['bp', 'bnf', 'b1950'])`, so I removed it as well.

CommonClimate commented:
Sorry if the back&forth makes things difficult for your test-writing ; I don't always see the whole picture when I'm writing the code, but this gave me the opportunity to think about it more, hence the changes.

kcpevey commented:
> I removed the if/then statement having to do with match_a, because we set exponent = 0 by default

The way its currently written (can't remember who coded this), it will warn if it doesn't match the input `time_unit` and then default to "years CE". Is that ok?

>Also, the case of "years BP" is already covered by elif any(c in tu for c in ['bp', 'bnf', 'b1950']), so I removed it as well.

So "years BP" must be input as time_unit of one of these: `['bp', 'bnf', 'b1950']`. These values `['yr BP', 'yrs BP', 'years BP']` will no longer be allowed to specify as time_unit. Is that ok?

CommonClimate commented:
> The way its currently written (can't remember who coded this), it will warn if it doesn't match the input time_unit and then default to "years CE". Is that ok?
Yes, this default makes sense for a lot of the examples we currently have. 

> So "years BP" must be input as time_unit of one of these: ['bp', 'bnf', 'b1950']. These values ['yr BP', 'yrs BP', 'years BP'] will no longer be allowed to specify as time_unit. Is that ok?

The way I wrote it, ['yr BP', 'yrs BP', 'years BP']  will be caught by the clause `elif any(c in tu for c in ['bp', 'bnf', 'b1950'])`, and assign the right `datum` and `direction`. Since `exponent` is set to 0 by default, we should be all good, no?

CommonClimate commented:
Hi @kcpevey , I starting debugging the merge from 'new_objects' into Development and found a few bugs in the way I cast the logical net over "time_units".  My latest commit (in a new branch, see https://github.com/LinkedEarth/Pyleoclim_util/commit/a1fa119437231fdd01bd39f3014aad7eb95a86aa) makes it explicit that the `time_unit` can contain compound statements like "ky BP", so the code now parses the first statement (about `exponent`) and then the second (BP, which is about `datum` and `direction`).  It's now passing all the pytests, some of which I modified to throw a few curve balls. Please have a look at your leisure.
